README.md
## rclone-mount for Android

![Hits](https://hitcounter.pythonanywhere.com/count/tag.svg?url=https%3A%2F%2Fgithub.com%2FMagisk-Modules-Repo%2Fcom.piyushgarg.rclone)
---

Mount cloud storage locally via rclone & fusermount directly on your Android powered smart device. 

Now you can have virtually limitless storage expansion with support for dozens of cloud providers. Extremely useful for devices without physical storage expansion capabilities & streaming large media files without need for full caching. Binaries obtained directly from rclone.org. 

---
## Features
- Support for arm, arm64, & x86

- Huge list of supported cloud storage providers

- Applications with ability to specify paths can access /mnt/cloud

- Most file explorers work just fine

- mpv (playstore) is best for playing audio and video files. Fast browsing and quick playing.

- Default dir caching set to 24h

- Executed mount scripts through other applications will be seen only by that application.

- Use adb shell only for executing mounting scripts so that all applications can see the mount points

- Mount points use names of remote(s) in rclone.conf

- Specify custom rclone params for each remote via `/sdcard/.rclone/.REMOTENAME.param`

## Configuration (post-installing)

1. Copy your rclone.conf file to `/sdcard/.rclone/rclone.conf`

2. Add custom params at `/sdcard/.rclone/.*.params`

3. Reboot or run `rclone-mount` from terminal

4. All your rclone mount points will show up under `/mnt/cloud/` & `/storage/cloud`

For more detailed configuration of rclone please refer to [official documentation](https://rclone.org)

## Custom Params

- Specification of rclone parameters on a per remote basis can be created in 

    `/sdcard/.rclone/.*.param`

   Where `*` is replace with name of remote



- List of available parameters and their default values:
  
      There is no need to specify values you do not wish to change.

        BUFFERSIZE=8M

        CACHEMAXSIZE=256M

        DIRCACHETIME=24h

        READAHEAD=128k

        CACHEMODE=writes

        CACHE=/data/rclone/cache

        CACHE_BACKEND=/data/rclone/cache-backend

## Known Issues

- VLC  takes a long time to load media as it opens file in write mode when using it's internal browser. 

    Create remote type alias for media dirs in rclone.conf and specify `CACHEMODE=off` in `/sdcard/.rclone/.ALIASNAME.param`

- Mount point can not be placed in `/sdcard/`

- Can not mount remotes until device is unlocked 

- Not responsilbe for any loss.

## Credits

- rclone devs
- pmj_pedro@xda
- rclone binaries from [rclone.org](https://rclone.org/downloads)
- fusermount binaries from  [xda-devs](https://forum.xda-developers.com/android/development/fusermount-android-rclone-mount-t3866652)

## Changelog

### v1.1
* Initial release
* rclone mount

### v1.2
* Change install process
* Changes for full systemless
* Improve mount reliability
* Symlink mountpoint to /storage/

### v1.3
* Move user rclone.conf & related to /sdcard/.rclone/
* Control global --vfs-cache-mode via simple files placed in /sdcard/.rclone/
* Specify custom params for individual remotes via /sdcard/.rclone/.REMOTENAME.params

## v1.4
* Update rclone to v1.57.0
* Fix paths


changelog.md
# Changelog
## v1.1
* Initial release
* rclone mount

## v1.2
* Change install process
* Changes for full systemless
* Improve mount reliability
* Link to /storage/

## v1.3
* Move user rclone.conf & related to /sdcard/.rclone/
* Control global --vfs-cache-mode via simple files placed in /sdcard/.rclone/
* Specify custom params for individual remotes via /sdcard/.rclone/.REMOTENAME.params

## v1.4
* Update rclone to v1.57.0
* Fix paths

install.sh
##########################################################################################
#
# Magisk Module Installer Script
#
##########################################################################################
##########################################################################################
#
# Instructions:
#
# 1. Place your files into system folder (delete the placeholder file)
# 2. Fill in your module's info into module.prop
# 3. Configure and implement callbacks in this file
# 4. If you need boot scripts, add them into common/post-fs-data.sh or common/service.sh
# 5. Add your additional or modified system properties into common/system.prop
#
##########################################################################################

##########################################################################################
# Config Flags
##########################################################################################

# Set to true if you do *NOT* want Magisk to mount
# any files for you. Most modules would NOT want
# to set this flag to true
SKIPMOUNT=false

# Set to true if you need to load system.prop
PROPFILE=false

# Set to true if you need post-fs-data script
POSTFSDATA=false

# Set to true if you need late_start service script
LATESTARTSERVICE=true

##########################################################################################
# Replace list
##########################################################################################

# List all directories you want to directly replace in the system
# Check the documentations for more info why you would need this

# Construct your list in the following format
# This is an example
REPLACE_EXAMPLE="
/system/app/Youtube
/system/priv-app/SystemUI
/system/priv-app/Settings
/system/framework
"

# Construct your own list here
REPLACE="
"

##########################################################################################
#
# Function Callbacks
#
# The following functions will be called by the installation framework.
# You do not have the ability to modify update-binary, the only way you can customize
# installation is through implementing these functions.
#
# When running your callbacks, the installation framework will make sure the Magisk
# internal busybox path is *PREPENDED* to PATH, so all common commands shall exist.
# Also, it will make sure /data, /system, and /vendor is properly mounted.
#
##########################################################################################
##########################################################################################
#
# The installation framework will export some variables and functions.
# You should use these variables and functions for installation.
#
# ! DO NOT use any Magisk internal paths as those are NOT public API.
# ! DO NOT use other functions in util_functions.sh as they are NOT public API.
# ! Non public APIs are not guranteed to maintain compatibility between releases.
#
# Available variables:
#
# MAGISK_VER (string): the version string of current installed Magisk
# MAGISK_VER_CODE (int): the version code of current installed Magisk
# BOOTMODE (bool): true if the module is currently installing in Magisk Manager
# MODPATH (path): the path where your module files should be installed
# TMPDIR (path): a place where you can temporarily store files
# ZIPFILE (path): your module's installation zip
# ARCH (string): the architecture of the device. Value is either arm, arm64, x86, or x64
# IS64BIT (bool): true if $ARCH is either arm64 or x64
# API (int): the API level (Android version) of the device
#
# Availible functions:
#
# ui_print <msg>
#     print <msg> to console
#     Avoid using 'echo' as it will not display in custom recovery's console
#
# abort <msg>
#     print error message <msg> to console and terminate installation
#     Avoid using 'exit' as it will skip the termination cleanup steps
#
# set_perm <target> <owner> <group> <permission> [context]
#     if [context] is empty, it will default to "u:object_r:system_file:s0"
#     this function is a shorthand for the following commands
#       chown owner.group target
#       chmod permission target
#       chcon context target
#
# set_perm_recursive <directory> <owner> <group> <dirpermission> <filepermission> [context]
#     if [context] is empty, it will default to "u:object_r:system_file:s0"
#     for all files in <directory>, it will call:
#       set_perm file owner group filepermission context
#     for all directories in <directory> (including itself), it will call:
#       set_perm dir owner group dirpermission context
#
##########################################################################################
##########################################################################################
# If you need boot scripts, DO NOT use general boot scripts (post-fs-data.d/service.d)
# ONLY use module scripts as it respects the module status (remove/disable) and is
# guaranteed to maintain the same behavior in future Magisk releases.
# Enable boot scripts by setting the flags in the config section above.
##########################################################################################

# Set what you want to display when installing your module

print_modname() {
  ui_print " "
  ui_print "*******************************"
  ui_print "*        rclone-mount         *"
  ui_print "*        Magisk Module        *"
  ui_print "*       by: piyushgarg        *"
  ui_print "*******************************"
  ui_print "*rclone: (v1.47.0) mod: (v1.3)*"
  ui_print "*******************************"
}

# Copy/extract your module files into $MODPATH in on_install.

on_install() {
  # The following is the default implementation: extract $ZIPFILE/system to $MODPATH
  # Extend/change the logic to whatever you want

  if [ "$ARCH" == "arm" ];then
    BINARY_PATH=$TMPDIR/binary/rclone-arm
  elif [ "$ARCH" == "arm64" ];then
    BINARY_PATH=$TMPDIR/binary/rclone-arm64
  elif [ "$ARCH" == "x86" ];then
    BINARY_PATH=$TMPDIR/binary/rclone-x86
  elif [ "$ARCH" == "x64" ];then
    BINARY_PATH=$TMPDIR/binary/rclone-x64
  fi

  CONFIG_PATH=$TMPDIR/config

  ui_print "* Detected arch: $ARCH"
  ui_print "+ Extracting package contents..."
  mkdir -p $MODPATH/system/bin
  ui_print "+ Extracting rclone-$ARCH"
  unzip -p "$ZIPFILE" binary/rclone-${ARCH} > $MODPATH/system/bin/rclone
  ui_print "+ Extracting fusermount-$ARCH"
  unzip -p "$ZIPFILE" binary/fusermount-${ARCH} > $MODPATH/system/bin/fusermount
  ui_print "+ Extracting rclone-mount script"
  unzip -p "$ZIPFILE" binary/rclone-mount > $MODPATH/system/bin/rclone-mount

}

# Only some special files require specific permissions
# This function will be called after on_install is done
# The default permissions should be good enough for most cases


set_permissions() {
  # The following is the default rule, DO NOT remove
  set_perm_recursive $MODPATH 0 0 0755 0644
  set_perm_recursive $MODPATH/system/bin 0 0 0755 0755
  set_perm $MODPATH/service.sh 0 0 0500
  # Here are some examples:
  # set_perm_recursive  $MODPATH/system/lib       0     0       0755      0644
  # set_perm  $MODPATH/system/bin/app_process32   0     2000    0755      u:object_r:zygote_exec:s0
  # set_perm  $MODPATH/system/bin/dex2oat         0     2000    0755      u:object_r:dex2oat_exec:s0
  # set_perm  $MODPATH/system/lib/libart.so       0     0       0644
}

# You can add more functions to assist your custom script code


common/manual.sh
#!/system/bin/sh
# Do NOT assume where your module will be located.
# ALWAYS use $MODDIR if you need to know where this script
# and module is placed.
# This will make sure your module will still work
# if Magisk change its mount point in the future

MODDIR=${0%/*}

. $MODDIR/module.prop


#RCLONE PARAMETERS
BUFFERSIZE=8M
CACHEMAXSIZE=256M
DIRCACHETIME=24h
READAHEAD=128k

USER_CONF=/sdcard/rclone.conf
CONFIGFILE=$MODDIR/rclone.conf
LOGFILE=/sdcard/rclone.log
HOME=/mnt
CLOUDROOTMOUNTPOINT=$HOME/cloud
CACHE=/mnt/runtime/default/rclone-cache
CACHE_BACKEND=/mnt/runtime/default/rc-cache-backend

if [[ ! -d $CLOUDROOTMOUNTPOINT ]]; then

mkdir -p $CLOUDROOTMOUNTPOINT

fi

if [[ ! -d $CACHE ]]; then

mkdir -p $CACHE

fi

if [[ ! -d $CACHE_BACKEND ]]; then

mkdir -p $CACHE_BACKEND

fi

ln -sf $CLOUDROOTMOUNTPOINT /mnt/runtime/read/cloud
ln -sf $CLOUDROOTMOUNTPOINT /mnt/runtime/write/cloud

if [[ -e $USER_CONFIG ]]; then

    cp $USER_CONFIG $CONFIGFILE
    chmod 0600 $CONFIGFILE
    
fi

#sh -c "$MODDIR/system/bin/rclone mount piyushDOTgarg_shopDOTmega: ${CLOUDROOTMOUNTPOINT}/piyushDOTgarg_shopDOTmega -vv --config ${CONFIGFILE} --attr-timeout 10m --cache-dir=/storage/cache --vfs-cache-mode writes --vfs-cache-max-age 168h0m0s --log-file /sdcard/dns.log --allow-other --gid 1015" &

rclone listremotes --config ${CONFIGFILE}|cut -f1 -d: |
        while read line; do
                echo "mounting... $line"
                mkdir -p ${CLOUDROOTMOUNTPOINT}/${line}
                rclone mount ${line}: ${CLOUDROOTMOUNTPOINT}/${line} --config ${CONFIGFILE} --max-read-ahead ${READAHEAD} --buffer-size ${BUFFERSIZE} --dir-cache-time ${DIRCACHETIME} --poll-interval 5m --attr-timeout ${DIRCACHETIME} --vfs-cache-mode writes --vfs-read-chunk-size 2M --vfs-read-chunk-size-limit 10M --vfs-cache-max-age 168h0m0s --vfs-cache-max-size ${CACHEMAXSIZE} --cache-dir=/storage/cache --cache-chunk-path /storage/cache-backend/ --cache-chunk-clean-interval 10m0s --log-file ${LOGFILE} --allow-other --gid 1015 --daemon
                sleep 5
        done

echo "...done"



common/service.sh
#!/system/bin/sh
# Do NOT assume where your module will be located.
# ALWAYS use $MODDIR if you need to know where this script
# and module is placed.
# This will make sure your module will work
# if Magisk changes it's mount point in the future

MODDIR=${0%/*}

#. $MODDIR/module.prop >> /dev/null 2>&1

id=com.piyushgarg.rclone

#MODULE VARS
USER_CONFDIR=/sdcard/.rclone
USER_CONF=$USER_CONFDIR/rclone.conf
CONFIGFILE=$MODDIR/rclone.conf
LOGFILE=/sdcard/rclone.log
HOME=/mnt
CLOUDROOTMOUNTPOINT=$HOME/cloud

#RCLONE PARAMETERS
BUFFERSIZE=8M
CACHEMAXSIZE=256M
DIRCACHETIME=24h
READAHEAD=128k
CACHEMODE=writes
CACHE=/data/rclone/cache
CACHE_BACKEND=/data/rclone/cache-backend

custom_params () {

    PARAMS="BUFFERSIZE CACHEMAXSIZE DIRCACHETIME READAHEAD CACHEMODE"
    BAD_SYNTAX="(^\s*#|^\s*$|^\s*[a-z_][^[:space:]]*=[^;&\(\`]*$)"

    if [[ -e $USER_CONFDIR/.$remote.param ]]; then 

        if ! egrep -q -iv "$BAD_SYNTAX" $USER_CONFDIR/.$remote.param; then

            for PARAM in ${PARAMS[@]}; do

                while read -r VAR; do

                    if [[ "$(echo "${VAR}" |grep -w "$PARAM")" ]]; then

                        eval $(echo "${VAR}" |cut -d ' ' -f 1)
 
                    fi

                done < $USER_CONFDIR/.$remote.param

            done

        else

            echo ".$remote.param contains bad syntax"

        fi

    fi

}

if [[ ! -d $CLOUDROOTMOUNTPOINT ]]; then

    mkdir -p $CLOUDROOTMOUNTPOINT

fi

if [[ ! -d $CACHE ]]; then

    mkdir -p $CACHE

fi

if [[ ! -d $CACHE_BACKEND ]]; then

    mkdir -p $CACHE_BACKEND

fi

if [[ ! -L /mnt/runtime/read/cloud ]]; then

    ln -sf $CLOUDROOTMOUNTPOINT /mnt/runtime/read/cloud
    
fi

if [[ ! -L /mnt/runtime/write/cloud ]]; then
    
    ln -sf $CLOUDROOTMOUNTPOINT /mnt/runtime/write/cloud
    
fi

until [[ $(getprop sys.boot_completed) = 1 ]] && [[ $(getprop dev.bootcomplete) = 1 ]] && [[ $(getprop service.bootanim.exit) = 1 ]] && [[ $(getprop init.svc.bootanim) = stopped ]] && [[ -e $USER_CONF ]] || [[ $COUNT -eq 240 ]]; do

    sleep 5
    ((++COUNT))

done

if [[ -e $USER_CONF ]]; then

    cp $USER_CONF $CONFIGFILE
    chmod 0600 $CONFIGFILE
    
fi

if [[ -e $USER_CONFDIR/.nocache ]]; then

    CACHEMODE=off
    
fi

if [[ -e $USER_CONFDIR/.mincache ]]; then

    CACHEMODE=minimal
    
fi

if [[ -e $USER_CONFDIR/.writecache ]]; then

    CACHEMODE=writes
    
fi

if [[ -e $USER_CONFDIR/.fullcache ]]; then

    CACHEMODE=full
    
fi

sleep 10

rclone listremotes --config ${CONFIGFILE}|cut -f1 -d: |
        while read line; do
                remote=$line
                custom_params
                echo "mounting... $remote"
                mkdir -p ${CLOUDROOTMOUNTPOINT}/${line}
                rclone mount ${remote}: ${CLOUDROOTMOUNTPOINT}/${remote} --config ${CONFIGFILE} --max-read-ahead ${READAHEAD} --buffer-size ${BUFFERSIZE} --dir-cache-time ${DIRCACHETIME} --poll-interval 5m --attr-timeout ${DIRCACHETIME} --vfs-cache-mode ${CACHEMODE} --vfs-read-chunk-size 2M --vfs-read-chunk-size-limit 10M --vfs-cache-max-age 10h0m0s --vfs-cache-max-size ${CACHEMAXSIZE} --cache-dir=${CACHE} --cache-chunk-path ${CACHE_BACKEND} --cache-chunk-clean-interval 10m0s --log-file ${LOGFILE} --allow-other --gid 1015 --daemon
                sleep 5
        done
echo
echo "...done"

